vm/src/stdlib/itertools.rs

```diff
@@ -181,6 +184,141 @@ impl PyItertoolsTakewhile {
     }
 }
 
+#[pyclass(name = "islice")]
+#[derive(Debug)]
+struct PyItertoolsIslice {
+    iterable: PyObjectRef,
+    cur: RefCell<usize>,
+    next: RefCell<usize>,
+    stop: Option<usize>,
+    step: usize,
+}
+
+impl PyValue for PyItertoolsIslice {
+    fn class(vm: &VirtualMachine) -> PyClassRef {
+        vm.class("itertools", "islice")
+    }
+}
+
+fn pyobject_to_usize(obj: PyObjectRef, vm: &VirtualMachine) -> Option<usize> {
+    let is_int = objtype::isinstance(&obj, &vm.ctx.int_type());
+    if is_int {
+        objint::get_value(&obj).to_usize()
+    } else {
+        None
+    }
+}
+
+#[pyimpl]
+impl PyItertoolsIslice {
+    #[pymethod(name = "__new__")]
+    fn new(_cls: PyClassRef, args: PyFuncArgs, vm: &VirtualMachine) -> PyResult {
+        let (iter, start, stop, step) = match args.args.len() {
+            0 | 1 => {
+                return Err(vm.new_type_error(format!(
+                    "islice expected at least 2 arguments, got {}",
+                    args.args.len()
+                )))
+            }
+
+            2 => {
```

Comment:
Argument binding is too complicated now, I want to improve this bit

Reply:
Maybe something like

```rust
// Get the argument indices
let (start, stop, step) = match args.args.len() {
  2 => (None, Some(1), None),
  3 => (Some(1), Some(2), None),
  4 => (Some(1), Some(2), Some(3)),
  0 | 1 => return Err("not enough args"),
  _ => return Err("too many args"),
};

// I don't know how you retrieve an argument by index.
let iter: PyObjectRef = get_arg(0).unwrap();
let start: usize = match start {
  Some(i) => get_arg(i).to_usize().ok_or_else(|| Err(...))?,
  _ => 0,
};
...
```

Reply:
Changed it so there's no code dup in the handling of `stop` argument.